Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -1133,11 +1141,31 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 {
     ASSERT(isLoaded());
 
+    BOOL uniqueIdentifierHasChanged = ![m_uniqueIdentifier isEqualToString:lastSeenUniqueIdentifier()];
+    if (uniqueIdentifierHasChanged)
+        saveUniqueIdentifierToStorage();
+
     if (!extension().hasBackgroundContent())
         return;
 
-    // FIXME: <https://webkit.org/b/246483> Handle non-persistent background pages differently here.
-    loadBackgroundWebView();
+    if (!isPersistent()) {
+        NSNumber *backgroundPageListenersVersionNumber = loadBackgroundPageListenersVersionNumberFromStorage();
+        BOOL savedVersionNumberDoesNotMatchCurrentVersionNumber = backgroundPageListenersVersionNumber.integerValue != currentBackgroundPageListenerStateVersion;
+
+        // FIXME: <https://webkit.org/b/248889> Check to see if background page listens to onStartup().
+        BOOL backgroundPageListensToOnStartup = NO;
+        EventListenterTypeCountedSet savedBackgroundPageListeners = loadBackgroundPageListenersFromStorage();
+
+        if (savedBackgroundPageListeners.isEmpty() || backgroundPageListensToOnStartup || savedVersionNumberDoesNotMatchCurrentVersionNumber || uniqueIdentifierHasChanged || (backgroundPageListensToOnStartup && m_isBeingLoadAsPartOfBrowserLaunch))
+            loadBackgroundWebView();
+        else {
+            m_backgroundPageListeners = savedBackgroundPageListeners;
```

Comment:
You can remove this once `loadBackgroundPageListenersFromStorage()` builds `m_backgroundPageListeners`.